### PR TITLE
DEVOP-617: org-wide go.mod replace-directive audit

### PR DESCRIPTION
**Pull request by srt0422**
## Summary

Adds a weekly + manual-dispatch audit that catches the Go-side
Shai-Hulud vector: a compromised `go.mod` `replace` directive
redirecting a legitimate import to an attacker fork. Complements
DEVOP-560 (PR #8) — DEVOP-560 is the deep daily forensic clone sweep,
this is the lighter no-clone Contents-API pass that runs weekly across
every Go module in the org.

Linear: https://linear.app/alloralabs/issue/DEVOP-617

## What this PR adds

1. **`.github/workflows/gomod-replace-audit.yml`** — weekly Mon 05:17 UTC + `workflow_dispatch`.
   - Enumerates every org Go module via `gh search code --owner allora-network 'filename:go.mod' --limit 200`.
   - Fetches each `go.mod` via `gh api repos/<r>/contents/<p>` (no cloning).
   - Runs the canonical awk extractor from `shai-hulud-defense/REFERENCE.md` (handles single-line and `replace (...)` block form).
   - Classifies each RHS:
     - `legitimate-allowlisted-host` — RHS host in canonical trusted-host allowlist (`github.com/(allora-network|cosmos|ethereum|fluxcd) | gopkg.in | google.golang.org | go.uber.org | go.opentelemetry.io | k8s.io | sigs.k8s.io`).
     - `legitimate-version-pin` — `LHS module path == RHS module path`. Structurally cannot redirect to an attacker fork; flagged only by the host filter otherwise.
     - `legitimate-local-relative` — `./...` / `../...` workspace replace.
     - `investigate-absolute` — `/...` (IOC-grade).
     - `SUSPICIOUS` — non-allowlisted host AND `LHS != RHS` (IOC-grade).
   - SUSPICIOUS or `investigate-absolute` → rolling GitHub Issue (label `gomod-replace-audit`, distinct from DEVOP-560's `shai-hulud-sweep`) + Slack page via `SLACK_SECURITY_WEBHOOK`.
   - Fetch failures → rolling-issue update only.
   - Permissions: `contents: read`, `issues: write`. All `uses:` SHA-pinned (matches DEVOP-560 pins).
   - `actionlint` clean (incl. shellcheck).

2. **`docs/security/gomod-replace-audit-2026-05-25.md`** — initial point-in-time audit report (executed locally before opening this PR).

3. **`docs/plans/2026-05-25-devop-617-gomod-replace-audit.md`** — short execution plan.

## Top-line audit results (from the report)

| Metric | Count |
|---|---|
| Go modules scanned | **12** (across 11 repos) |
| Modules with zero replace directives | **8** |
| Total `replace` directives found | **6** |
| Allowlisted-host RHS | 2 |
| Same-path version-pin RHS (non-allowlisted host but `LHS == RHS`) | 4 |
| Local relative / absolute / **SUSPICIOUS** | **0 / 0 / 0** |
| **Escalated to incident response** | **0** |

**No SUSPICIOUS findings. No escalation.** Every current `replace` directive in the org is either:
- on an allowlisted host (cosmos), OR
- a same-path version pin (`LHS module path == RHS module path` — structurally cannot redirect).

Specifically the four non-allowlisted entries are all same-path version pins:
- `allora-chain/go.mod`: `gin-gonic/gin v1.9.1` and `syndtr/goleveldb v1.0.1-...` (canonical Cosmos SDK simapp pins; adjacent comments reference `cosmos/cosmos-sdk#10409`).
- `allora-sdk-go/go.mod` + `forge-v2/backend/go.mod`: `cometbft/cometbft v0.38.17` (canonical Cosmos BFT consensus engine).

See [the full audit report](./docs/security/gomod-replace-audit-2026-05-25.md) for the row-by-row classification + recommendations (adding `github.com/cometbft` to the canonical allowlist is the obvious follow-up).

## Coordination with PR #8 (DEVOP-560)

- Different branch (`scott/devop-617-gomod-replace-audit`), different files (workflow + plan + report all new).
- Different rolling-issue label (`gomod-replace-audit` vs `shai-hulud-sweep`) so the two pipelines don't collide.
- Cron offset (Mon 05:17 UTC vs daily 04:07 UTC) so they don't compete for the same scheduler slot.
- This workflow does NOT touch `scripts/shai-hulud-ioc-sweep.sh` or any of PR #8's files.

## Test plan

- [x] `actionlint .github/workflows/gomod-replace-audit.yml` (incl. shellcheck) — clean.
- [x] Local end-to-end audit produced 6 findings, 0 SUSPICIOUS (matches the report).
- [ ] `workflow_dispatch` after merge to validate the workflow run produces an artifact + (with current org state) leaves the rolling issue untouched.
- [ ] Inject a synthetic suspicious replace in a sandbox repo and re-run to confirm the rolling-issue + Slack path fires.

Made with [Cursor](https://cursor.com)

---
## Summary by cubic
Adds a weekly and manual org audit that checks every `go.mod` `replace` for attacker redirects and alerts on anything outside the trusted-host allowlist. Meets DEVOP-617 acceptance criteria; improves parsing, scope checks (including partial private-coverage detection), and alerting to avoid false-clean runs and noisy pages.

- **New Features**
  - Adds `.github/workflows/gomod-replace-audit.yml` (Mon 05:17 UTC + `workflow_dispatch`).
  - Scans via `gh search code` + `gh api` (no cloning); classifies replaces; `SUSPICIOUS` or absolute → rolling issue + Slack; fetch failures → issue only.
  - Uploads `audit.tsv` and `summary.md`; SHA-pinned actions; minimal perms (`contents: read`, `issues: write`).
  - Token-scope probe: fail loudly if the token can’t see private repos; require `secrets.GH_ORG_READ_TOKEN` or ack public-only runs via `vars.ACCEPT_PUBLIC_ONLY_AUDIT=true`.
  - Docs added (plan + initial report). Initial audit: 12 modules, 6 replaces, 0 suspicious.

- **Bug Fixes**
  - AWK extractor: skip full-line `//` comments and strip trailing `// ...` on real directives to avoid false SUSPICIOUS.
  - Contents API: normalize JSON null to empty; accept spec-valid `go.mod` with a `module` directive anywhere (leading comments allowed); otherwise treat as fetch failure.
  - Slack paging: gate on `success()` and non-empty outputs; final run summary reflects audit failure vs clean.
  - Warn when `gh search code` hits the 200-result cap.
  - Corrected report: zero-replace modules is 9; added “modules with at least one replace” row.
  - Scope probe: paginate private-repo listing to detect partial coverage (selected-repo tokens); fail unless acknowledged via `vars.ACCEPT_PUBLIC_ONLY_AUDIT=true`.

<sup>Written for commit be4718d7c3d76de5480af72c7b4ba0c7583b5d0e. Summary will update on new commits. <a href="https://cubic.dev/pr/allora-network/.github/pull/9?utm_source=github">Review in cubic</a></sup>

### CODE REVIEW

File: .github/workflows/gomod-replace-audit.yml [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+          #   3. total_private > 0 AND        -> SCOPE GAP, fail unless ack'd
+          #      visible_private == 0
+          if [ "$total_private" = "0" ]; then
+            echo "::notice::Org has no private repos; public-only audit covers full scope."
+            exit 0
+          fi
+          if [ "$visible_private" -gt 0 ] 2>/dev/null; then
+            echo "::notice::Token has org-read scope; audit will cover both public and private Go modules."
+            exit 0
+          fi
```
[the remark this answers is not in the document]
Comment (srt0422) [reply .github/workflows/gomod-replace-audit.yml:140]: > This probe treats visibility of one private repo as full private coverage, so selected-repository PAT/App tokens can still leave some private repos unaudited while the workflow reports complete coverage.

Addressed in commit `aeb5cc0`. The probe now paginates the full `orgs/$ORG/repos?type=private` list (via `gh api --paginate ... --jq '.[].id' | wc -l`) and compares the visible count against `total_private_repos` from the org object. Three new branches:

- `total_private == 0` → public-only org, clean audit (unchanged)
- `total_private` known AND `visible_private == total_private` → full coverage, full audit
- `total_private == "unknown"` (org-object permission denied) AND `visible_private > 0` → at least some private visibility, proceed best-effort with a warning
- otherwise → SCOPE GAP, fail (or warn-only if `ACCEPT_PUBLIC_ONLY_AUDIT=true`)

A selected-repository PAT with access to a single private repo now correctly classifies as SCOPE GAP and refuses to run a partial audit.

**Comment by srt0422** [comment]
**Needs-human follow-up #3 verification done.**

Re-checked the close-out concern that \`aeb5cc0\` swept up the 'Cross-pipeline follow-ups (deferred)' section in \`docs/security/gomod-replace-audit-2026-05-25.md\`:

**State now:** the deletion was REVERSED by \`be4718d\` (2026-05-26 13:09 PT, after the close-out at 10:26 PT). The deferred-follow-ups section is back in the doc (lines ~102\u2013129 on the branch HEAD). No action needed in the PR itself.

**Linear tracking:** the three items in the deferred section had no dedicated Linear coverage prior to this run \u2014 they were tracked only in the at-risk doc. Filed three Low-priority follow-up tickets so the tracking is durable beyond the doc:

| Item | Ticket |
|---|---|
| Extract \`GO_TRUSTED_HOSTS_RE\` to a single committed source (so workflow + sweep + REFERENCE.md don't drift) | [DEVOP-632](https://linear.app/alloralabs/issue/DEVOP-632) |
| Fixture-based parser tests for go.mod \`replace\` extraction (8+ edge cases + classification matrix) | [DEVOP-633](https://linear.app/alloralabs/issue/DEVOP-633) |
| Regex lookalike-bypass corpus test for the \`(/\|$)\` boundary anchor in \`GO_TRUSTED_HOSTS_RE\` | [DEVOP-634](https://linear.app/alloralabs/issue/DEVOP-634) |

(Note: the deferred section also lists a fourth item \u2014 'extract the awk replace-directive extractor to scripts/extract-go-replace.awk' \u2014 which I folded into DEVOP-632's acceptance criteria since extracting the shared regex and extracting the shared extractor are the same atomic 'consolidate the duplicated parsing pipeline' chunk of work. Happy to split it out if anyone disagrees.)

All three tickets link back to this PR and to the specific anchor line in the deferred section.

**Comment by srt0422** [comment]
## Regression discovered during review-fix-loop on HEAD `be4718d`

**Severity: P1 (functional security regression — re-opens the original `cubic-dev-ai` finding from `PRRT_kwDOLZ5Xss6EqQn3`).**

The token-scope probe in `be4718d` ("fix(workflows): allow valid go.mod with leading comments to pass sanity check") **inadvertently reverted** the paginate-and-count fix that landed in `aeb5cc0`. The current HEAD probe is back to:

```yaml
visible_private=$(gh api "orgs/$ORG/repos?type=private&per_page=1" --jq 'length' 2>/dev/null || echo 0)
...
if [ "$visible_private" -gt 0 ] 2>/dev/null; then
  echo "::notice::Token has org-read scope; audit will cover both public and private Go modules."
  exit 0
fi
```

This is exactly the original cubic concern: a **selected-repository PAT or GitHub App token** that grants access to a single private repo will produce `visible_private = 1`, the `> 0` branch fires, and the audit proceeds while silently false-cleaning every other private Go module in the org. `aeb5cc0` correctly fixed this with a paginated `--paginate ... --jq '.[].id' | wc -l` count compared against `total_private_repos` (require equality, four-case matrix). That logic must be restored.

The cubic-dev-ai inline comment thread on `.yml:140` shows "✅ Addressed in `aeb5cc0`" because cubic's resolution detector evaluates the commit it landed in, not the latest HEAD — so the GREEN status check on this PR is misleading for this specific finding on the post-`aeb5cc0` branch tip.

### Other findings on this PR (verified status on HEAD `be4718d`)

| Finding | Severity | Cubic-marked addressed in | Verified on HEAD | Action |
|---|---|---|---|---|
| Token fallback false-clean (`yml:62`) | P1 | `1d05588` | Mitigated by scope probe — but mitigation depends on a working probe (see above) | Re-fix probe |
| `gomod-replace-audit-2026-05-25.md:30` data inconsistency | P2 | `1d05588` | RESOLVED (line 30 reads "9", list has 9 entries) | none |
| Partial private coverage probe (`yml:140`) | P1 | `aeb5cc0` | **REGRESSED in `be4718d`** | **fix below** |
| go.mod sanity check rejects leading comments (`yml:222`) | P2 | `aeb5cc0` | RESOLVED (`grep -qm1 '^[[:space:]]*module '`) | none |
| Grammar in `gomod-replace-audit-2026-05-25.md:104` | P3 | `aeb5cc0` | non-blocking | none (P3 policy) |

### Proposed fix

Restore the `aeb5cc0` probe block verbatim onto HEAD while keeping every other `be4718d` improvement intact (sanity-check via `grep -qm1`, `.content // ""` JSON-null normalization, awk comment-line skip + trailing-comment strip, Slack `success()` guard, final-summary `AUDIT_OUTCOME` branch). Patch attached below — it cleanly applies on top of `be4718d`.

### Why this comment instead of a push

The `review-fix-loop` runner attempted to land the fix as a follow-up commit (`b42d884` locally), but the available `gh` OAuth tokens (`gist, read:org, repo`) lack the `workflow` scope required by GitHub to update `.github/workflows/*.yml` paths. A maintainer with a `workflow`-scoped token (or the `cursor[bot]` / `srt0422` GitHub App with workflow scope) needs to land the commit directly, or this PR can be re-run by an autonomous loop with a workflow-scoped token.

### Patch (apply on top of `be4718d`)

```diff
diff --git a/.github/workflows/gomod-replace-audit.yml b/.github/workflows/gomod-replace-audit.yml
index 2a0a773..4042c42 100644
--- a/.github/workflows/gomod-replace-audit.yml
+++ b/.github/workflows/gomod-replace-audit.yml
@@ -117,35 +117,50 @@ jobs:
           if [ -z "$total_private" ]; then
             total_private="unknown"
           fi
-          # Listing returns the COUNT of private repos visible to the
-          # token. Compare against `total_private_repos` to detect a
-          # token scope gap.
-          visible_private=$(gh api "orgs/$ORG/repos?type=private&per_page=1" \
-            --jq 'length' 2>/dev/null || echo 0)
+          # Paginate ALL private repos visible to the token and count them.
+          # Earlier versions used `per_page=1 | length` which only proved
+          # *at least one* private repo was visible — a selected-repository
+          # PAT or GitHub App token granting access to a single private repo
+          # would falsely report "full coverage" while leaving every other
+          # private repo unaudited. See PR #9 cubic finding PRRT_kwDOLZ5Xss6EqQn3.
+          # `--paginate` walks all pages; `.[].id` flattens each page; wc -l
+          # gives the total count.
+          visible_private=$(gh api --paginate "orgs/$ORG/repos?type=private&per_page=100" \
+            --jq '.[].id' 2>/dev/null | wc -l | tr -d ' ')
+          # Defensive empty/non-numeric guard (gh failure paths can yield "").
+          if ! printf '%s' "$visible_private" | grep -qE '^[0-9]+$'; then
+            visible_private=0
+          fi
           echo "Org private-repo count: $total_private"
           echo "Private repos visible to token: $visible_private"
 
-          # Three cases:
-          #   1. total_private == 0           -> public-only org, clean audit
-          #   2. visible_private > 0          -> token has org scope, full audit
-          #   3. total_private > 0 AND        -> SCOPE GAP, fail unless ack'd
-          #      visible_private == 0
+          # Four cases (treat partial visibility as a scope gap — selected-
+          # repository PAT/App tokens are the original cubic concern):
+          #   1. total_private == 0                         -> public-only org, clean audit
+          #   2. total_private known AND counts equal       -> full private coverage, full audit
+          #   3. total_private == "unknown" AND visible > 0 -> at least some private visibility; accept (org-object permission denied)
+          #   4. otherwise                                  -> SCOPE GAP, fail unless ack'd
           if [ "$total_private" = "0" ]; then
             echo "::notice::Org has no private repos; public-only audit covers full scope."
             exit 0
           fi
-          if [ "$visible_private" -gt 0 ] 2>/dev/null; then
-            echo "::notice::Token has org-read scope; audit will cover both public and private Go modules."
+          if [ "$total_private" != "unknown" ] \
+              && [ "$visible_private" -eq "$total_private" ] 2>/dev/null; then
+            echo "::notice::Token has full private-repo visibility ($visible_private/$total_private); audit will cover both public and private Go modules."
+            exit 0
+          fi
+          if [ "$total_private" = "unknown" ] && [ "$visible_private" -gt 0 ] 2>/dev/null; then
+            echo "::warning::Cannot read orgs/$ORG to confirm total_private_repos; token has access to $visible_private private repos but coverage completeness cannot be verified. Proceeding (best-effort)."
             exit 0
           fi
           # Scope gap. Honor the operator-ack'd public-only mode if set.
           if [ "${ACCEPT_PUBLIC_ONLY_AUDIT:-false}" = "true" ]; then
-            echo "::warning::Token lacks private-repo visibility (ACCEPT_PUBLIC_ONLY_AUDIT=true). Auditing PUBLIC repos only — private Go modules will not be scanned."
+            echo "::warning::Token has partial private-repo visibility (visible=$visible_private, total_private_repos=$total_private; ACCEPT_PUBLIC_ONLY_AUDIT=true). Auditing PUBLIC + visible-private repos only — non-visible private Go modules will not be scanned."
             exit 0
           fi
           # Failure path: emit an actionable error and abort.
-          echo "::error::Token cannot see private repos in $ORG (total_private_repos=$total_private, visible=$visible_private). This audit would silently false-clean every private Go module."
-          echo "::error::Fix one of: (a) provision GH_ORG_READ_TOKEN secret with read:org + repo:read; or (b) set ACCEPT_PUBLIC_ONLY_AUDIT=true at org-vars level to consciously accept a public-only audit."
+          echo "::error::Token has partial private-repo visibility in $ORG (visible=$visible_private, total_private_repos=$total_private). This audit would silently false-clean $(($total_private - $visible_private)) private Go modules."
+          echo "::error::Fix one of: (a) provision GH_ORG_READ_TOKEN secret with read:org + repo:read (org-wide); or (b) set ACCEPT_PUBLIC_ONLY_AUDIT=true at org-vars level to consciously accept a partial-coverage audit."
           exit 1
```

### Recommended PR-side guardrail

Once this fix lands, consider adding a tiny shell unit-test fixture under `scripts/test-fixtures/gomod-replace-audit-probe/` that replays the `total_private` × `visible_private` matrix against the embedded probe (extracted to a separate sourceable script) so a future "small fix" can't silently revert the four-case logic again. That extraction is already on the deferred-follow-ups list (DEVOP-633 / DEVOP-634) so it's a natural fit.

File: .github/workflows/gomod-replace-audit.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (srt0422) [reply .github/workflows/gomod-replace-audit.yml:140]: Re-flagging this thread as **NOT resolved on current HEAD `be4718d`**, contrary to the cubic auto-marker.

The paginate-and-count probe that landed in `aeb5cc0` was inadvertently reverted in `be4718d` ("fix(workflows): allow valid go.mod with leading comments to pass sanity check"). The current HEAD probe is back to `gh api orgs/$ORG/repos?type=private&per_page=1 --jq 'length'` with a `> 0` accept branch — the exact pattern the original cubic finding flagged.

Full writeup, severity matrix, and the diff that needs to land are in [the new top-level PR comment](https://github.com/allora-network/.github/pull/9#issuecomment-4628925508). The fix is committed locally (`b42d884` on the worktree) but cannot be pushed by the review-fix-loop runner because its `gh` token is missing the `workflow` OAuth scope; a maintainer with workflow scope needs to land it.

Marking this PR `blocked` for the autonomous loop until the patch is applied.

File: .github/workflows/gomod-replace-audit.yml [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+          set -euo pipefail
+          mkdir -p "$OUTPUT_DIR"
+          # `gh search code` returns repo + file path tuples. Some repos
+          # have multiple go.mod files (monorepo sub-modules); the dedup
+          # happens on the `repo<TAB>path` pair, not on `repo` alone.
+          # The code-search API hard-caps at 1000 results; we gate the
+          # overflow check on the RAW (pre-dedup) hit count so `sort -u`
+          # cannot mask API-level truncation that would produce a
+          # false-clean audit report.
+          gh search code --owner "$ORG" 'filename:go.mod' --limit 1000 \
```
[the remark this answers is not in the document]
Comment (srt0422) [reply .github/workflows/gomod-replace-audit.yml]: > A timed-out code search with fewer than 1000 hits still passes this gate and can produce a false-clean audit, because `incomplete_results` is discarded.

Addressed in `91f13d4`. Enumeration is no longer a single org-wide `gh search code` call; it is sharded per repository:

- Repos are listed via `gh api "orgs/$ORG/repos?per_page=100" --paginate` (fully paginated, no result ceiling; forks excluded for parity with code-search defaults).
- Each shard runs `gh api "search/code?q=repo:<repo>+filename:go.mod&per_page=100"`, which exposes the envelope `gh search code --json` hides (confirmed: `incomplete_results` is not an available `--json` field).
- `incomplete_results == true` on any shard aborts the workflow with `::error::` (fail closed) naming the shard.
- A shard `total_count > 100` (single-page overflow; `per_page=100` is the search-API max) also aborts loudly, with shard pagination as the documented follow-up — this replaces the prior 1000-result gate, since sharding keeps every query far under the ceiling that REST pagination cannot pass.
- Shard calls are paced at 2.5s (code search is 30 req/min authenticated) with bounded retry (3 attempts, 5s/15s/45s backoff); a persistent shard failure aborts the audit rather than producing a partial enumeration.

Verified: YAML parses, `actionlint` clean, downstream `paths_file`/`output_dir`/`count` output contract unchanged.

**Comment by srt0422** [comment]
**Final cross-provider review** (agent-code-review, 3-pass)
verdict: Ready with fixes
counts: P0=0 P1=0 P2=2 P3=1
coverage: Pass 1 — correctness+security: kimi-k3 orchestrator inline (user-mandated substitution for Claude Opus 4.8; claude-code CLI not invoked); style: GLM 5.2 via OpenRouter (substituted for Claude Sonnet 5). Pass 2 — correctness: google/gemini-3.1-pro-preview; security: deepseek/deepseek-v4-pro; style: meta-llama/llama-4-maverick. Pass 3 — correctness: meta-llama/llama-4-maverick; security: GLM 5.2 (substituted for Claude Sonnet 5); style: meta-llama/llama-4-scout. All non-orchestrator passes via OpenRouter HTTP, temperature 0.1. Pass 2/3 findings source-verified by the orchestrator before emission.

| severity | file:line | category | reviewer | summary | suggested_fix |
| --- | --- | --- | --- | --- | --- |
| P2 | .github/workflows/gomod-replace-audit.yml:289 | correctness | gemini-3.1-pro (pass 2), source-verified | The sanity grep `^[[:space:]]*module ` requires a literal space after `module`, so a valid tab-separated directive (`module\texample.com/foo`, legal per the modfile grammar) is misrouted to fetch-failures and the module's replace directives go unaudited with only operational-grade visibility. | Accept any whitespace after the keyword: `grep -qm1 '^[[:space:]]*module[[:space:]]' "$gomod"`. |
| P2 | .github/workflows/gomod-replace-audit.yml:336 | correctness | gemini-3.1-pro (pass 2), source-verified | The local-relative case pattern `./*|../*` does not match the exact paths `.` or `..`, so a legal bare local replace (`replace x => ..`) falls through to SUSPICIOUS and fires a false incident-grade page. | Extend the pattern to `.|..|./*|../*)`. |
| P3 | .github/workflows/gomod-replace-audit.yml:183 | style | kimi-k3 (pass 1) | The `shard` counter is incremented per loop iteration but never read (the final summary echoes `$repo_count` instead). | Remove `shard=0` and the `shard=$((shard + 1))` increment, or use it in the per-shard log lines. |

Refuted (source-verified false positives, not findings):

| claim | pass | refutation |
| --- | --- | --- |
| `SUSPICIOUS` is inconsistently capitalized vs the lowercase classes and complicates case-sensitive filtering (rename to lowercase) | GLM 5.2 (pass 1, style) | Intentional alarm salience: `SUSPICIOUS` is the only uppercase class and doubles as the incident-grade operator term used consistently in the summary heading, rolling issue, and Slack page; case-sensitive filtering is unambiguous precisely because it is the sole uppercase value. |

Blocking P2 rows above are recorded as `final_review_residual` by review-fix-loop (Step E does not auto-fix; re-running the loop picks them up as fix-now).
